docs/development/out-of-tree.md

```diff
@@ -0,0 +1,96 @@
+(out-of-tree)=
+
+# Building and testing Python packages out of Tree
+
+This is some information about how to build and test Python packages against
+Pyodide out of tree (for instance in your package's CI or for use with private
+packages).
+
+## Building binary packages for Pyodide
+
+If your package is a pure Python package (i.e., if the wheel ends in
+`py3-none-any.whl`) then it can be built in whichever way you normally build it.
+Otherwise, the procedure is simple. In your package directory run the following
+command line commands:
+
+```sh
+pip install pyodide-build`
+pyodide build`
+```
+
+Pyodide currently only supports Linux for out of tree builds, though there is a
+good change it will work in MacOS too. If you are using Windows, try Windows
+Subsystem Linux.
+
+`pyodide build` invokes a slightly modified version of the `pypa/build` build
+frontend so the behavior should be similar to what happens if you do:
+
+```sh
+pip install build
+python -m build
+```
+
+If you run into problems, make sure that building a native wheel with
+`pypa/build` works.
+
+`pyodide build` respects the environment variables `CFLAGS`, `CXXFLAGS`, and
+`LDFLAGS`, so if you need to customize compiler arguments you can set these. Any
+additional arguments passed to `pyodide build` will be passed along to the build
+backend.
+
+If you run into problems, please open an issue about it.
+
+## Testing packages against Pyodide
+
+Pyodide provides an experimental command line runner for testing packages
+against Pyodide. The way it works is simple: you can create a virtual
+environment with:
+
+```sh
+pyodide venv .venv-pyodide
+```
+
+Enter it just like a normal virtual environment:
+
+```sh
+source .venv-pyodide/bin/python
```

Comment:
Maybe add a note that it requires Node?

Reply:
What ever happened to the `node-fetch` requirement? Do people not need that anymore?

Reply:
What do you mean? I think we are using it for node < 18 which doesn't have builtin fetch.

https://github.com/pyodide/pyodide/blob/0eb6036b8556800eb0d074e2f9c08db8a32f7a35/src/js/compat.ts#L40

Reply:
But say you have node 14 installed on your system and you do `pip install pyodide-build` and then start using the command line runner. How does `node-fetch` get installed?

Reply:
It is not installed by `pip install pyodide-build` (obviously). Probably you installed it globally before, or probably you did `make` before. During make we install `node-fetch` "locally" while installing Pyodide JS package, and then we copy `node_modules` into the Pyodide root (which is quite strange and I think we can remove it now :/), so `node-fetch` is available when you are in Pyodide root repository.

https://github.com/pyodide/pyodide/blob/0eb6036b8556800eb0d074e2f9c08db8a32f7a35/Makefile#L70-L73

Reply:
Okay great. And the reason that it works in CI is that I am testing the out of tree build stuff in circleci where node-fetch is already installed. Perhaps we should try to install `node-fetch` as part of creating the virtualenv because other than node-fetch it's self contained.

Reply:
Sounds great. It would be great if Node >= v18 becomes dominant so we can simply tell people to use Node v18 and drop all node-fetch stuff... (The active LTS of Node v18 starts next month)

Reply:
Wow node-fetch v3 has 8mb of `node_modules`, I guess because it requires web-streams-polyfill which is 200kb of code but it has 15 different variants and each variant has a source map. But node-fetch v2 does not require web-streams-polyfill and leads to only 512kb of `node_modules`.

Reply:
Okay, I updated `create_xbuildenv` to vendor the appropriate `node-fetch` stuff into `node_modules`. This should mean that we don't need any comment about `node-fetch`. It's also good to have the `node_modules` folder in the xbuildenv root because that means that the execution environment won't be polluted by some random other `node_modules` folder in a parent directory.